# AG-HYPOPT · Trial 03

> ▶ RUN = execute this cell, never modify it  ·  ✍️ WRITE = replace the [placeholder] with your text  ·  ✍️ SET = put your value where the [placeholder] is, then run the cell. Full rules: `instructions.md`


This is trial_03 of experiment_2 (cap 20), the third of the five uniform warm-up draws. The registry holds two completed trials: trial_01 (objective 0.0016, current best) and trial_02 (objective 0.0537, run with n_iter=10 as a mechanism test). Setup unchanged: 8 synthetic experiments (1nW and 3nW at Trans05, Trans20, Trans60, Trans100), targets at the true values with fixed seeds; four tunables (sigma_ref, lr_mu, lr_gamma, gamma_anneal). Note: n_iter is currently 10 (mechanism smoke), so results are not directly comparable to trial_01.


In [ ]:
# 2. ▶ RUN - Propose candidate trials (AGHyperopt)
TRIAL_ID = 'trial_03'   # auto-stamped by the generator; do not edit
import os

EXPERIMENT_DIR = os.getcwd()
SPACE_PATH = os.path.join(EXPERIMENT_DIR, 'space.json')
TRIALS_PATH = os.path.join(EXPERIMENT_DIR, 'trials.json')

MAX_TRIALS = 20    # campaign cap: stop generating new trials after this many (None = unlimited)

from ag_hypopt import AGHyperopt

opt = AGHyperopt()
opt.fit(SPACE_PATH, TRIALS_PATH)
proposed_trials = opt.propose_trials(10)


phase = uniform  (2/5 trials)
ID |     ei | params
 1 |      - | {"gamma_anneal": 0.5804670364169725, "lr_gamma": 0.5511027518016418, "lr_mu": 26.464947997784563, "sigma_ref": 18.947360581187276}
 2 |      - | {"gamma_anneal": 0.07063301091573715, "lr_gamma": 0.9804978813094047, "lr_mu": 24.028492549758823, "sigma_ref": 20.721286105539075}
 3 |      - | {"gamma_anneal": 0.0960852245066594, "lr_gamma": 0.5603087503164537, "lr_mu": 14.26995060581453, "sigma_ref": 23.535299776972035}
 4 |      - | {"gamma_anneal": 0.48289884006049844, "lr_gamma": 0.8582092906166641, "lr_mu": 16.085354970683277, "sigma_ref": 9.544774435695537}
 5 |      - | {"gamma_anneal": 0.4159385902618761, "lr_gamma": 0.2510538048833403, "lr_mu": 25.69077929981455, "sigma_ref": 17.633287982441296}
 6 |      - | {"gamma_anneal": 0.5685658050640303, "lr_gamma": 0.48362077450389473, "lr_mu": 29.267450609872583, "sigma_ref": 22.862422426443953}
 7 |      - | {"gamma_anneal": 0.5837876228053214, "lr_gamma": 0.35571096628157

The registry holds two completed trials: trial_01 (objective 0.0016) and trial_02 (objective 0.0537, n_iter=10 mechanism test). Phase: uniform warm-up (3/5). Note: in the uniform phase the proposal RNG is seeded, so this candidate table is identical to trial_02's; candidate 9 (= trial_02's chosen config) is therefore excluded to avoid repeating a configuration. Among the remaining candidates I take the most central draw by scaled distance to the range midpoints: candidate 4 (sigma_ref 9.54, lr_mu 16.09, lr_gamma 0.858, gamma_anneal 0.483). I choose candidate 4 (4 goes into the SET cell below).


In [ ]:
INDEX = 4            # candidate 4 (most central untried uniform draw)


In [ ]:
# 5. ▶ RUN - Execute the chosen trial  (do not interrupt unless obviously broken)
assert INDEX is not None, 'INDEX not set: edit the SET cell (cell 4) before running this one'
assert 1 <= INDEX <= len(proposed_trials), f'bad INDEX: {INDEX}'
CHOSEN = proposed_trials[INDEX - 1]['params']
print('chosen:', CHOSEN)

from ag_hypopt import objective

import traceback, time
t0 = time.time()
try:
    loss, uncertainty, report = objective(CHOSEN)
    print(f'trial finished in {(time.time()-t0)/60:.1f} min')
    print(f'objective = {loss:.4f} ± {uncertainty:.4f}')
    print(report)
except Exception:
    traceback.print_exc()
    loss, uncertainty = None, None


chosen: {'gamma_anneal': 0.48289884006049844, 'lr_gamma': 0.8582092906166641, 'lr_mu': 16.085354970683277, 'sigma_ref': 9.544774435695537}
Running  1nW Trans05 ... 

μ 9.39 -> 5.36 | γ 8.5 -> 7.53 | NLL 1.32


Running  1nW Trans20 ... 

μ 17.32 -> 11.84 | γ 8.5 -> 10.29 | NLL 3.38


Running  1nW Trans60 ... 

μ 61.37 -> 41.78 | γ 8.5 -> 11.29 | NLL 5.78


Running 1nW Trans100 ... 

μ 70.82 -> 51.81 | γ 8.5 -> 9.61 | NLL 4.14


Running  3nW Trans05 ... 

μ 13.20 -> 7.46 | γ 14.1 -> 11.49 | NLL 1.71


Running  3nW Trans20 ... 

μ 34.28 -> 24.71 | γ 14.1 -> 15.57 | NLL 3.41


Running  3nW Trans60 ... 

μ 103.20 -> 77.05 | γ 14.1 -> 15.21 | NLL 3.98


Running 3nW Trans100 ... 

μ 175.71 -> 123.42 | γ 14.1 -> 14.66 | NLL 3.38



Total: 6.4 min
trial finished in 6.4 min
objective = 0.0696 ± 0.0146
exp           μ_true   μ_fit     Δμ% |   γ_true   γ_fit     Δγ%
1nW Trans05     9.39    5.36   -42.9 |     8.50    7.53   -11.4
1nW Trans20    17.32   11.84   -31.6 |     8.50   10.29   +21.1
1nW Trans60    61.37   41.78   -31.9 |     8.50   11.29   +32.9
1nW Trans100   70.82   51.81   -26.8 |     8.50    9.61   +13.0
3nW Trans05    13.20    7.46   -43.5 |    14.10   11.49   -18.5
3nW Trans20    34.28   24.71   -27.9 |    14.10   15.57   +10.4
3nW Trans60   103.20   77.05   -25.3 |    14.10   15.21    +7.9
3nW Trans100  175.71  123.42   -29.8 |    14.10   14.66    +4.0

μ: RMSE 23.274 (rel 33.1%, bias -17.732) | γ: RMSE 1.720 (rel 17.1%, bias +0.657)


The trial ran the 8-experiment benchmark in 6.4 minutes with config (sigma_ref 9.54, lr_mu 16.09, lr_gamma 0.858, gamma_anneal 0.483; n_runs = 100, n_iter = 10). Measured objective 0.0696 +/- 0.0146. mu RMSE 23.27 (relative 33.1 percent, bias -17.73); gamma RMSE 1.72 (relative 17.1 percent, bias +0.66). mu lands low on all eight experiments (-25 to -44 percent); gamma is scattered (-18 to +33 percent). The registry now holds trial_01 (0.0016, current best), trial_02 (0.0537) and this trial_03 (0.0696); n_iter=10 smoke, so none are comparable to trial_01.

**Summary:** trial_03 (uniform warm-up draw: sigma_ref 9.54, lr_mu 16.09, lr_gamma 0.858, gamma_anneal 0.483; n_runs=100, n_iter=10) ran the 8-experiment benchmark in 6.4 min: objective 0.0696 +/- 0.0146, mu rel-RMSE 33.1 percent, gamma rel-RMSE 17.1 percent.
**Key insight:** mu is biased low on every experiment at n_iter=10; the campaign registry best remains trial_01.

[write-up written by the agent (Pukky).]


In [ ]:
SUMMARY = "trial_03 (uniform warm-up draw: sigma_ref 9.54, lr_mu 16.09, lr_gamma 0.858, gamma_anneal 0.483; n_runs=100, n_iter=10) ran the 8-experiment benchmark in 6.4 min: objective 0.0696 +/- 0.0146, mu rel-RMSE 33.1 percent, gamma rel-RMSE 17.1 percent."
KEY_INSIGHT = "mu is biased low on every experiment at n_iter=10; registry best remains trial_01."


In [ ]:
# 8. ▶ RUN - Record the trial in trials.json
import json

assert SUMMARY is not None, 'SUMMARY not set: fill the SET cell (cell 7) first'
assert KEY_INSIGHT is not None, 'KEY_INSIGHT not set: fill the SET cell (cell 7) first'

entry = {
    'trial_id': TRIAL_ID,
    'config': CHOSEN,
    'objective': loss,
    'uncertainty': uncertainty,
    'summary': SUMMARY,
    'key_insight': KEY_INSIGHT,
    'notebook': f'{TRIAL_ID}.ipynb',
}
data = json.load(open(TRIALS_PATH))
assert not any(t.get('trial_id') == TRIAL_ID for t in data['trials']),     f'{TRIAL_ID} is already registered in trials.json'
data['trials'].append(entry)
json.dump(data, open(TRIALS_PATH, 'w'), indent=2)
best = min((t for t in data['trials'] if t.get('objective') is not None),
           key=lambda t: t['objective'], default=None)
print('saved', TRIAL_ID, '| current best:', best['trial_id'] if best else None)


saved trial_03 | current best: trial_01


In [ ]:
# 9. ▶ RUN - Generate the next trial (or end the campaign)
import os, re, json as _json

num = int(re.search(r'(\d+)$', TRIAL_ID).group(1))
nxt = num + 1
if MAX_TRIALS is not None and nxt > MAX_TRIALS:
    print(f'Campaign complete: cap MAX_TRIALS={MAX_TRIALS} reached after {TRIAL_ID}. Stop here.')
else:
    template_path = os.path.join(EXPERIMENT_DIR, 'template.ipynb')
    target = os.path.join(EXPERIMENT_DIR, f'trial_{nxt:02d}.ipynb')
    assert not os.path.exists(target), f'{target} already exists: refusing to overwrite'

    nb = _json.load(open(template_path))

    def _stamp(old, new):
        for c in nb['cells']:
            if old in ''.join(c.get('source', [])):
                c['source'] = [s.replace(old, new) for s in c['source']]
                return True
        raise RuntimeError(f'stamp target {old!r} not found in the template')

    _stamp('{N}', f'{nxt:02d}')
    _stamp('trial_XXX', f'trial_{nxt:02d}')
    _json.dump(nb, open(target, 'w'), indent=1)
    print(f'Created {os.path.basename(target)}. Open it and follow its cells from the top.')


Created trial_04.ipynb. Open it and follow its cells from the top.
